examples/exex/op-proposer/src/op_proposer.rs

```
@@ -0,0 +1,195 @@
+    T: Transport + Clone,
+    N: Network,
+    P: Provider<T, N>,
+{
+    let l1_block = provider
+        .get_block(BlockId::from(BlockNumberOrTag::Latest), false)
+        .await?
+        .ok_or(eyre!("L1 block not found"))?;
+
+    let l1_block_hash = l1_block.header.hash.ok_or(eyre!("L1 Block hash not found"))?;
+    let l1_block_number = l1_block.header.number.ok_or(eyre!("L1 block number not found"))?;
+
+    Ok(L1BlockAttributes { hash: l1_block_hash, number: l1_block_number })
+}
+
+async fn get_l2_safe_head<T, N, P>(provider: Arc<P>) -> eyre::Result<u64>
+where
+    T: Transport + Clone,
+    N: Network,
+    P: Provider<T, N>,
+{
+    let sync_status: OptimismSyncStatus =
+        provider.client().request("optimism_syncStatus", ()).await?;
+    let safe_head = sync_status.safe_l2.number;
+    Ok(safe_head)
+}
+
+pub struct OpProposer<T, N, P>
+where
+    T: Transport + Clone,
+    N: Network,
+    P: Provider<T, N> + 'static,
+{
+    l1_provider: Arc<P>,
+    rollup_provider: String,
+    l2_to_l1_message_passer: Address,
+    _pd: PhantomData<fn() -> (T, N)>,
+}
+
+impl<T, N, P> OpProposer<T, N, P>
+where
+    T: Transport + Clone,
+    N: Network,
+    P: Provider<T, N> + 'static,
+{
+    pub fn new(
+        l1_provider: Arc<P>,
+        rollup_provider: String,
+        l2_to_l1_message_passer: Address,
+    ) -> Self {
+        Self { l1_provider, rollup_provider, l2_to_l1_message_passer, _pd: PhantomData }
+    }
+
+    pub async fn run<Node: FullNodeComponents>(
+        &self,
+        mut ctx: ExExContext<Node>,
+        mut l2_output_db: L2OutputDb,
+        l2_output_oracle: Arc<L2OutputOracleInstance<T, Arc<P>, N>>,
+        mut transaction_manager: TxManager<T, N, P>,
+    ) -> eyre::Result<()> {
+        let l2_provider = ctx.provider().clone();
+        let rollup_provider = Arc::new(ProviderBuilder::new().with_recommended_fillers().on_http(
+            self.rollup_provider.parse().expect("Could not parse rollup provider endpoint"),
```

this should return an error